codex/stores/repostore.nim

```diff
@@ -40,127 +42,268 @@ declareGauge(codex_repostore_bytes_used, "codex repostore bytes used")
+  minusRefCount: Natural = 0
+): Future[?!void] {.async.} =
+  if cid.isEmpty:
+    return success()
+
+  without metaKey =? createBlockExpirationMetadataKey(cid), err:
     return failure(err)
 
-  without value =? await self.metaDs.get(key), err:
-    if err of DatastoreKeyNotFound:
-      trace "Cid not found", treeCid, index
-      return failure(newException(BlockNotFoundError, err.msg))
-    else:
-      trace "Error getting cid from datastore", err = err.msg, key
-      return failure(err)
+  await self.metaDs.modifyT(metaKey,
+    proc (maybeCurrBlockMd: ?BlockMetadata): Future[?BlockMetadata] {.async.} =
+      if currBlockMd =? maybeCurrBlockMd:
+        BlockMetadata(
+          size: currBlockMd.size,
+          refCount: currBlockMd.refCount + plusRefCount - minusRefCount
+        ).some
+      else:
+        raise newException(BlockNotFoundError, "Metadata for block with cid " & $cid & " not found")
+  )
+
+proc storeBlock(self: RepoStore, blk: Block): Future[?!StoreResult] {.async.} =
+  if blk.isEmpty:
+    return success(StoreResult(kind: AlreadyInStore))
+
+  without metaKey =? createBlockExpirationMetadataKey(blk.cid), err:
+    return failure(err)
+
+  without blkKey =? makePrefixKey(self.postFixLen, blk.cid), err:
+    return failure(err)
+
+  await self.metaDs.modifyTGetU(metaKey,
+    proc (maybeCurrMd: ?BlockMetadata): Future[(?BlockMetadata, StoreResult)] {.async.} =
+      var
+        md: BlockMetadata
+        res: StoreResult
+
+      if currMd =? maybeCurrMd:
+        if currMd.size == blk.data.len.NBytes:
+          md = BlockMetadata(size: currMd.size, refCount: currMd.refCount)
+          res = StoreResult(kind: AlreadyInStore)
+
+          # making sure that the block acutally is stored in the repoDs
+          without hasBlock =? await self.repoDs.has(blkKey), err:
+            raise err
+
+          if not hasBlock:
+            warn "Block metadata is present, but block is absent. Restoring block.", cid = blk.cid
+            if err =? (await self.repoDs.put(blkKey, blk.data)).errorOption:
+              raise err
+        else:
+          raise newException(CatchableError, "Repo already stores a block with the same cid but with a different size, cid: " & $blk.cid)
+      else:
+        md = BlockMetadata(size: blk.data.len.NBytes, refCount: 0)
+        res = StoreResult(kind: Stored, used: blk.data.len.NBytes)
+        if err =? (await self.repoDs.put(blkKey, blk.data)).errorOption:
+          raise err
+
+      (md.some, res)
+  )
+
+proc tryDeleteBlock(self: RepoStore, cid: Cid): Future[?!DeleteResult] {.async.} =
+  if cid.isEmpty:
+    return success(DeleteResult(kind: InUse))
+
+  without metaKey =? createBlockExpirationMetadataKey(cid), err:
+    return failure(err)
+
+  without blkKey =? makePrefixKey(self.postFixLen, cid), err:
+    return failure(err)
+
+  await self.metaDs.modifyTGetU(metaKey,
+    proc (maybeCurrMd: ?BlockMetadata): Future[(?BlockMetadata, DeleteResult)] {.async.} =
```

Comment:
Perhaps this would be a bit more readable if it was broken up:
1. get metadata from datastore
2. try to delete the block
3. update the metadata in datastore

Reply:
All these things are happening here. The whole point is it has to be done atomically, hence `modifyGet` has to be used.